File: CLI Pulse Bar/CLIPulseCore/Sources/CLIPulseCore/PDFReportGenerator.swift
#if canImport(PDFKit) && !os(watchOS)
import Foundation
import PDFKit
import os
#if canImport(AppKit)
import AppKit
private typealias PlatformColor = NSColor
private typealias PlatformFont = NSFont
#elseif canImport(UIKit)
import UIKit
private typealias PlatformColor = UIColor
private typealias PlatformFont = UIFont
#endif

private let pdfLogger = Logger(subsystem: "com.clipulse", category: "PDFReportGenerator")

/// Generates a monthly usage/cost report as a PDF document.
public enum PDFReportGenerator {

    /// Resolves the default user-visible save destination for an exported
    /// PDF. iter22: previously dumped into `temporaryDirectory`, which
    /// vanished from Finder/Downloads. macOS now prefers `~/Downloads`;
    /// iOS keeps temp because Downloads isn't a thing there and the
    /// share sheet handles user routing.
    ///
    /// `now` and `existing` are injected so tests can reproduce
    /// collisions without touching the real filesystem.
    public static func defaultDestination(
        for date: Date,
        existing: (URL) -> Bool = { FileManager.default.fileExists(atPath: $0.path) }
    ) -> (preferred: URL, fallback: URL) {
        let baseName = "cli-pulse-report-\(dateString(date))"
        let temp = FileManager.default.temporaryDirectory
            .appendingPathComponent("\(UUID().uuidString)-\(baseName).pdf")
        #if canImport(AppKit)
        let downloads = FileManager.default.urls(for: .downloadsDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: "\(realUserHome())/Downloads", isDirectory: true)
        let candidate = uniqueDestination(in: downloads, baseName: baseName, ext: "pdf", existing: existing)
        return (preferred: candidate, fallback: temp)
        #else
        return (preferred: temp, fallback: temp)
        #endif
    }

    /// Pure helper for `defaultDestination` — picks
    /// `<base>.pdf`, `<base>-2.pdf`, `<base>-3.pdf`, … until one
    /// doesn't already exist (per the injected `existing`).
    static func uniqueDestination(
        in directory: URL,
        baseName: String,
        ext: String,
        existing: (URL) -> Bool
    ) -> URL {
        var candidate = directory.appendingPathComponent("\(baseName).\(ext)")
        var n = 2
        while existing(candidate) {
            candidate = directory.appendingPathComponent("\(baseName)-\(n).\(ext)")
            n += 1
        }
        return candidate
    }

    // MARK: - Public API

    /// Generate a monthly PDF report and write to disk.
    ///
    /// iter23: when `destinationURL` is provided (e.g. selected by
    /// the user via `NSSavePanel`), write there directly — sandbox
    /// grants write access transitively through the panel's
    /// security-scoped URL. When `nil`, fall back to the
    /// `defaultDestination(...)` resolver (Downloads → temp).
    /// Returns the file URL on success, nil on failure.
    public static func generateReport(
        dashboard: DashboardSummary?,
        providers: [ProviderUsage],
        sessions: [SessionRecord],
        dailyUsage: [DailyUsage],
        costForecast: CostForecast?,
        generatedDate: Date = Date(),
        destinationURL: URL? = nil
    ) -> URL? {
        let pageWidth: CGFloat = 612  // US Letter
        let pageHeight: CGFloat = 792
        let margin: CGFloat = 50
        let contentWidth = pageWidth - margin * 2

        let pdfData = NSMutableData()
        var mediaBox = CGRect(x: 0, y: 0, width: pageWidth, height: pageHeight)

        guard let consumer = CGDataConsumer(data: pdfData as CFMutableData),
              let context = CGContext(consumer: consumer, mediaBox: &mediaBox, nil) else {
            return nil
        }

        var y: CGFloat = pageHeight - margin

        func newPage() {
            if y < pageHeight - margin { // Don't end a page we haven't drawn on
                context.endPage()
            }
            context.beginPage(mediaBox: &mediaBox)
            y = pageHeight - margin
        }

        func checkSpace(_ needed: CGFloat) {
            if y - needed < margin {
                newPage()
            }
        }

        // ── Page 1: Header + Summary ──
        context.beginPage(mediaBox: &mediaBox)

        // Title
        y = drawText(L10n.pdf.title, at: CGPoint(x: margin, y: y), fontSize: 22, bold: true, context: context)
        y -= 4

        // iter22: respect the in-app language override when one is
        // active so dates render in the chosen locale; otherwise
        // follow the system. The lproj name (e.g. "ja", "zh-Hans")
        // maps cleanly onto a `Locale` identifier.
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .long
        if let override = LocaleOverrideStore.shared.override {
            dateFormatter.locale = Locale(identifier: override)
        }
        y = drawText(L10n.pdf.generated(dateFormatter.string(from: generatedDate)), at: CGPoint(x: margin, y: y), fontSize: 10, color: .gray, context: context)
        y -= 20

        // Divider
        y = drawDivider(at: y, x: margin, width: contentWidth, context: context)
        y -= 12

        // Summary section
        if let d = dashboard {
            y = drawText(L10n.pdf.summary, at: CGPoint(x: margin, y: y), fontSize: 16, bold: true, context: context)
            y -= 8

            let summaryItems: [(String, String)] = [
                (L10n.pdf.todayUsage, formatTokens(d.total_usage_today)),
                (L10n.pdf.todayEstimatedCost, CostFormatter.format(d.total_estimated_cost_today)),
                (L10n.pdf.activeSessions, "\(d.active_sessions)"),
                (L10n.pdf.onlineDevices, "\(d.online_devices)"),
                (L10n.pdf.unresolvedAlerts, "\(d.unresolved_alerts)"),
            ]

            for (label, value) in summaryItems {
                y = drawKeyValue(label, value: value, at: y, x: margin, width: contentWidth, fontSize: 11, context: context)
            }
            y -= 12
        }

        // Cost Forecast section
        if let forecast = costForecast, forecast.isReliable {
            y = drawDivider(at: y, x: margin, width: contentWidth, context: context)
            y -= 8
            y = drawText(L10n.pdf.costForecast, at: CGPoint(x: margin, y: y), fontSize: 16, bold: true, context: context)
            y -= 8

            y = drawKeyValue(L10n.pdf.monthEndEstimate, value: CostFormatter.format(forecast.predictedMonthTotal), at: y, x: margin, width: contentWidth, fontSize: 11, context: context)
            y = drawKeyValue(L10n.pdf.spentSoFar, value: CostFormatter.format(forecast.actualToDate), at: y, x: margin, width: contentWidth, fontSize: 11, context: context)
            y = drawKeyValue(L10n.pdf.confidenceRange, value: "\(CostFormatter.format(forecast.lowerBound)) — \(CostFormatter.format(forecast.upperBound))", at: y, x: margin, width: contentWidth, fontSize: 11, context: context)
            y = drawKeyValue(L10n.pdf.progress, value: L10n.pdf.progressValue(forecast.currentDayOfMonth, forecast.daysInMonth), at: y, x: margin, width: contentWidth, fontSize: 11, context: context)
            y -= 12
        }

        // Provider breakdown
        y = drawDivider(at: y, x: margin, width: contentWidth, context: context)
        y -= 8
        y = drawText(L10n.pdf.providerBreakdown, at: CGPoint(x: margin, y: y), fontSize: 16, bold: true, context: context)
        y -= 8

        // Table header
        let colWidths: [CGFloat] = [contentWidth * 0.3, contentWidth * 0.2, contentWidth * 0.2, contentWidth * 0.15, contentWidth * 0.15]
        let headers = [L10n.pdf.hProvider, L10n.pdf.hWeekUsage, L10n.pdf.hEstCost, L10n.pdf.hRemaining, L10n.pdf.hQuota]
        y = drawTableRow(headers, at: y, x: margin, colWidths: colWidths, fontSize: 9, bold: true, context: context)

        y = drawDivider(at: y, x: margin, width: contentWidth, context: context, thin: true)

        let sortedProviders = providers.sorted { $0.estimated_cost_week > $1.estimated_cost_week }
        for p in sortedProviders {
            checkSpace(16)
            let row = [
                p.provider,
                formatTokens(p.week_usage),
                CostFormatter.format(p.estimated_cost_week),
                p.remaining.map { formatTokens($0) } ?? L10n.pdf.na,
                p.quota.map { formatTokens($0) } ?? L10n.pdf.na,
            ]
            y = drawTableRow(row, at: y, x: margin, colWidths: colWidths, fontSize: 9, context: context)
        }
        y -= 12

        // Top sessions by cost
        checkSpace(40)
        y = drawDivider(at: y, x: margin, width: contentWidth, context: context)
        y -= 8
        y = drawText(L10n.pdf.topSessions, at: CGPoint(x: margin, y: y), fontSize: 16, bold: true, context: context)
        y -= 8

        let sessionColWidths: [CGFloat] = [contentWidth * 0.25, contentWidth * 0.25, contentWidth * 0.15, contentWidth * 0.2, contentWidth * 0.15]
        let sessionHeaders = [L10n.pdf.hProvider, L10n.pdf.hProject, L10n.pdf.hCost, L10n.pdf.hUsage, L10n.pdf.hStatus]
        y = drawTableRow(sessionHeaders, at: y, x: margin, colWidths: sessionColWidths, fontSize: 9, bold: true, context: context)
        y = drawDivider(at: y, x: margin, width: contentWidth, context: context, thin: true)

        let topSessions = sessions.sorted { $0.estimated_cost > $1.estimated_cost }.prefix(15)
        for s in topSessions {
            checkSpace(16)
            let row = [
                s.provider,
                s.project,
                CostFormatter.format(s.estimated_cost),
                formatTokens(s.total_usage),
                s.status,
            ]
            y = drawTableRow(row, at: y, x: margin, colWidths: sessionColWidths, fontSize: 9, context: context)
        }
        y -= 12

        // Daily cost trend (text-based)
        if !dailyUsage.isEmpty {
            checkSpace(40)
            y = drawDivider(at: y, x: margin, width: contentWidth, context: context)
            y -= 8
            y = drawText(L10n.pdf.dailyTrend, at: CGPoint(x: margin, y: y), fontSize: 16, bold: true, context: context)
            y -= 8

            let costByDate = dailyUsage.reduce(into: [String: Double]()) { result, entry in
                result[entry.date, default: 0] += entry.cost
            }
            let sortedDates = costByDate.keys.sorted().suffix(30)
            let maxCost = costByDate.values.max() ?? 1.0

            for date in sortedDates {
                checkSpace(14)
                let cost = costByDate[date] ?? 0
                let barWidth = maxCost > 0 ? CGFloat(cost / maxCost) * (contentWidth - 130) : 0

                // Date label
                _ = drawText(String(date.suffix(5)), at: CGPoint(x: margin, y: y), fontSize: 8, color: .gray, context: context)

                // Bar
                context.setFillColor(CGColor(red: 0.2, green: 0.5, blue: 1.0, alpha: 0.7))
                context.fill(CGRect(x: margin + 45, y: y - 2, width: barWidth, height: 8))

                // Cost label
                _ = drawText(CostFormatter.format(cost), at: CGPoint(x: margin + 50 + (contentWidth - 130), y: y), fontSize: 8, context: context)
                y -= 12
            }
        }

        // Footer
        checkSpace(30)
        y -= 10
        y = drawDivider(at: y, x: margin, width: contentWidth, context: context)
        y -= 4
        // iter22: pull the version from the bundle so the footer
        // doesn't rot when we ship new builds. Falls back to the
        // current iter version (1.15.0) if Info.plist isn't present
        // (e.g. running this generator outside an app bundle in
        // tests / SPM contexts).
        let appVersion = (Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String) ?? "1.15.0"
        _ = drawText(L10n.pdf.footer(appVersion, dateFormatter.string(from: generatedDate)), at: CGPoint(x: margin, y: y), fontSize: 8, color: .gray, context: context)

        context.endPage()
        context.closePDF()

        // iter23: prefer the caller-supplied destination
        // (NSSavePanel-selected URL on macOS). Falls back to
        // `defaultDestination` (Downloads → temp) when nil — used by
        // tests, iOS share-sheet callers, and any path that can't
        // present a save panel.
        if let destinationURL {
            do {
                try (pdfData as Data).write(to: destinationURL)
                return destinationURL
            } catch {
                pdfLogger.warning("PDF export to user-selected URL \(destinationURL.path, privacy: .public) failed: \(error.localizedDescription, privacy: .public)")
                return nil
            }
        }
        // iter22 default: `~/Downloads/cli-pulse-report-YYYY-MM-DD.pdf`
        // (or a `-2`, `-3`… suffix on collision). Fall back to temp
        // if Downloads write fails.
        let dest = defaultDestination(for: generatedDate)
        do {
            try (pdfData as Data).write(to: dest.preferred)
            return dest.preferred
        } catch {
            pdfLogger.warning("PDF export to \(dest.preferred.path, privacy: .public) failed: \(error.localizedDescription, privacy: .public). Falling back to temp.")
            do {
                try (pdfData as Data).write(to: dest.fallback)
                return dest.fallback
            } catch {
                pdfLogger.error("PDF export fallback also failed: \(error.localizedDescription, privacy: .public)")
                return nil
            }
        }
    }

    // MARK: - Drawing Helpers

    private static func drawText(
        _ text: String,
        at point: CGPoint,
        fontSize: CGFloat,
        bold: Bool = false,
        color: PlatformColor = PlatformColor(red: 0.1, green: 0.1, blue: 0.1, alpha: 1),
        context: CGContext
    ) -> CGFloat {
        let font: PlatformFont
        if bold {
            font = PlatformFont.boldSystemFont(ofSize: fontSize)
        } else {
            font = PlatformFont.systemFont(ofSize: fontSize)
        }
        let attrs: [NSAttributedString.Key: Any] = [
            .font: font,
            .foregroundColor: color,
        ]
        let attrString = NSAttributedString(string: text, attributes: attrs)
        let size = attrString.size()
        let drawPoint = CGPoint(x: point.x, y: point.y - size.height)

        #if canImport(AppKit)
        NSGraphicsContext.saveGraphicsState()
        NSGraphicsContext.current = NSGraphicsContext(cgContext: context, flipped: false)
        attrString.draw(at: drawPoint)
        NSGraphicsContext.restoreGraphicsState()
        #elseif canImport(UIKit)
        UIGraphicsPushContext(context)
        // UIKit draws top-down; we need to flip for our coordinate system
        context.saveGState()
        context.translateBy(x: 0, y: point.y)
        context.scaleBy(x: 1, y: -1)
        attrString.draw(at: CGPoint(x: point.x, y: 0))
        context.restoreGState()
        UIGraphicsPopContext()
        #endif

        return point.y - size.height - 2
    }

    private static func drawKeyValue(
        _ key: String,
        value: String,
        at y: CGFloat,
        x: CGFloat,
        width: CGFloat,
        fontSize: CGFloat,
        context: CGContext
    ) -> CGFloat {
        _ = drawText(key, at: CGPoint(x: x, y: y), fontSize: fontSize, color: .gray, context: context)
        _ = drawText(value, at: CGPoint(x: x + width * 0.5, y: y), fontSize: fontSize, bold: true, context: context)
        return y - fontSize - 6
    }

    private static func drawTableRow(
        _ values: [String],
        at y: CGFloat,
        x: CGFloat,
        colWidths: [CGFloat],
        fontSize: CGFloat,
        bold: Bool = false,
        context: CGContext
    ) -> CGFloat {
        var offsetX = x
        for (i, value) in values.enumerated() {
            let w = i < colWidths.count ? colWidths[i] : 80
            // Truncate if too long
            let truncated = value.count > 25 ? String(value.prefix(22)) + "..." : value
            _ = drawText(truncated, at: CGPoint(x: offsetX, y: y), fontSize: fontSize, bold: bold, context: context)
            offsetX += w
        }
        return y - fontSize - 5
    }

    private static func drawDivider(
        at y: CGFloat,
        x: CGFloat,
        width: CGFloat,
        context: CGContext,
        thin: Bool = false
    ) -> CGFloat {
        context.setStrokeColor(CGColor(gray: 0.8, alpha: 1))
        context.setLineWidth(thin ? 0.5 : 1.0)
        context.move(to: CGPoint(x: x, y: y))
        context.addLine(to: CGPoint(x: x + width, y: y))
        context.strokePath()
        return y - 2
    }

    // MARK: - Utilities

    private static func formatTokens(_ value: Int) -> String {
        if value >= 1_000_000 {
            return String(format: "%.1fM", Double(value) / 1_000_000)
        } else if value >= 1_000 {
            return String(format: "%.1fK", Double(value) / 1_000)
        }
        return "\(value)"
    }

    private static func dateString(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: date)
    }
}
#endif

File: android/app/src/main/java/com/clipulse/android/util/PdfReportGenerator.kt
package com.clipulse.android.util

import android.content.Context
import android.graphics.Canvas
import com.clipulse.android.R
import android.graphics.Color
import android.graphics.Paint
import android.graphics.pdf.PdfDocument
import com.clipulse.android.BuildConfig
import com.clipulse.android.data.model.CostForecast
import com.clipulse.android.data.model.DailyUsage
import com.clipulse.android.data.model.DashboardSummary
import com.clipulse.android.data.model.ProviderUsage
import com.clipulse.android.data.model.SessionRecord
import java.io.File
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import com.clipulse.android.ui.common.DataTokenDisplay
import com.clipulse.android.ui.common.DateDisplay
import android.text.format.DateFormat

object PdfReportGenerator {

    private const val PAGE_WIDTH = 612
    private const val PAGE_HEIGHT = 792
    private const val MARGIN = 50f
    private val CONTENT_WIDTH = PAGE_WIDTH - MARGIN * 2

    fun generate(
        context: Context,
        dashboard: DashboardSummary?,
        providers: List<ProviderUsage>,
        sessions: List<SessionRecord>,
        dailyUsage: List<DailyUsage>,
        costForecast: CostForecast?,
    ): File? {
        val doc = PdfDocument()
        var pageNum = 1
        var pageInfo = PdfDocument.PageInfo.Builder(PAGE_WIDTH, PAGE_HEIGHT, pageNum).create()
        var page = doc.startPage(pageInfo)
        var canvas = page.canvas
        var y = MARGIN

        val titlePaint = Paint().apply { textSize = 22f; isFakeBoldText = true; color = Color.parseColor("#1a1a1a") }
        val headingPaint = Paint().apply { textSize = 16f; isFakeBoldText = true; color = Color.parseColor("#1a1a1a") }
        val bodyPaint = Paint().apply { textSize = 11f; color = Color.parseColor("#333333") }
        val labelPaint = Paint().apply { textSize = 10f; color = Color.GRAY }
        val boldPaint = Paint().apply { textSize = 11f; isFakeBoldText = true; color = Color.parseColor("#1a1a1a") }
        val smallPaint = Paint().apply { textSize = 9f; color = Color.parseColor("#333333") }
        val smallBoldPaint = Paint().apply { textSize = 9f; isFakeBoldText = true; color = Color.parseColor("#1a1a1a") }
        val linePaint = Paint().apply { color = Color.parseColor("#cccccc"); strokeWidth = 1f }
        val thinLinePaint = Paint().apply { color = Color.parseColor("#dddddd"); strokeWidth = 0.5f }
        val barPaint = Paint().apply { color = Color.parseColor("#3380FF"); alpha = 180 }

        fun newPage() {
            doc.finishPage(page)
            pageNum++
            pageInfo = PdfDocument.PageInfo.Builder(PAGE_WIDTH, PAGE_HEIGHT, pageNum).create()
            page = doc.startPage(pageInfo)
            canvas = page.canvas
            y = MARGIN
        }

        fun checkSpace(needed: Float) {
            if (y + needed > PAGE_HEIGHT - MARGIN) newPage()
        }

        fun divider(thin: Boolean = false) {
            canvas.drawLine(MARGIN, y, MARGIN + CONTENT_WIDTH, y, if (thin) thinLinePaint else linePaint)
            y += 4
        }

        val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.US)
        // The report's language: Context resources follow Android's per-app language.
        val locale = context.resources.configuration.locales[0]
        val today = DateDisplay.longDate(java.time.LocalDate.now(), locale)
        val monthDayPattern = DateFormat.getBestDateTimePattern(locale, "Md")

        // Title
        canvas.drawText(context.getString(R.string.pdf_title), MARGIN, y + 22f, titlePaint)
        y += 30f
        canvas.drawText(context.getString(R.string.pdf_generated, today), MARGIN, y + 10f, labelPaint)
        y += 20f
        divider()
        y += 8f

        // Summary
        if (dashboard != null) {
            canvas.drawText(context.getString(R.string.pdf_summary), MARGIN, y + 16f, headingPaint)
            y += 24f

            fun kv(label: String, value: String) {
                canvas.drawText(label, MARGIN, y + 11f, labelPaint)
                canvas.drawText(value, MARGIN + CONTENT_WIDTH * 0.5f, y + 11f, boldPaint)
                y += 16f
            }

            kv(context.getString(R.string.pdf_today_usage), formatTokens(dashboard.totalUsageToday))
            kv(context.getString(R.string.pdf_today_est_cost), "$${String.format("%.2f", dashboard.totalEstimatedCostToday)}")
            kv(context.getString(R.string.pdf_active_sessions), "${dashboard.activeSessions}")
            kv(context.getString(R.string.pdf_online_devices), "${dashboard.onlineDevices}")
            kv(context.getString(R.string.pdf_unresolved_alerts), "${dashboard.unresolvedAlerts}")
            y += 8f
        }

        // Forecast
        if (costForecast != null && costForecast.isReliable) {
            divider()
            y += 4f
            canvas.drawText(context.getString(R.string.pdf_cost_forecast), MARGIN, y + 16f, headingPaint)
            y += 24f

            fun kv(label: String, value: String) {
                canvas.drawText(label, MARGIN, y + 11f, labelPaint)
                canvas.drawText(value, MARGIN + CONTENT_WIDTH * 0.5f, y + 11f, boldPaint)
                y += 16f
            }

            kv(context.getString(R.string.pdf_month_end_estimate), "$${String.format("%.2f", costForecast.predictedMonthTotal)}")
            kv(context.getString(R.string.pdf_spent_so_far), "$${String.format("%.2f", costForecast.actualToDate)}")
            kv(context.getString(R.string.pdf_confidence_range), "$${String.format("%.2f", costForecast.lowerBound)} — $${String.format("%.2f", costForecast.upperBound)}")
            kv(context.getString(R.string.pdf_progress), context.getString(R.string.pdf_progress_value, costForecast.currentDayOfMonth, costForecast.daysInMonth))
            y += 8f
        }

        // Provider breakdown
        divider()
        y += 4f
        canvas.drawText(context.getString(R.string.pdf_provider_breakdown), MARGIN, y + 16f, headingPaint)
        y += 24f

        // Table header
        val provCols = floatArrayOf(0f, 0.3f, 0.5f, 0.7f, 0.85f)
        val provHeaders = arrayOf(context.getString(R.string.pdf_h_provider), context.getString(R.string.pdf_h_week_usage), context.getString(R.string.pdf_h_est_cost), context.getString(R.string.pdf_h_remaining), context.getString(R.string.pdf_h_quota))
        for (i in provHeaders.indices) {
            canvas.drawText(provHeaders[i], MARGIN + CONTENT_WIDTH * provCols[i], y + 9f, smallBoldPaint)
        }
        y += 14f
        divider(thin = true)

        val sortedProviders = providers.sortedByDescending { it.estimatedCostWeek }
        for (p in sortedProviders) {
            checkSpace(14f)
            val row = arrayOf(
                p.provider.take(20),
                formatTokens(p.weekUsage),
                "$${String.format("%.2f", p.estimatedCostWeek)}",
                p.remaining?.let { formatTokens(it) } ?: context.getString(R.string.pdf_na),
                p.quota?.let { formatTokens(it) } ?: context.getString(R.string.pdf_na),
            )
            for (i in row.indices) {
                canvas.drawText(row[i], MARGIN + CONTENT_WIDTH * provCols[i], y + 9f, smallPaint)
            }
            y += 13f
        }
        y += 8f

        // Top sessions
        checkSpace(40f)
        divider()
        y += 4f
        canvas.drawText(context.getString(R.string.pdf_top_sessions), MARGIN, y + 16f, headingPaint)
        y += 24f

        val sessCols = floatArrayOf(0f, 0.25f, 0.5f, 0.65f, 0.85f)
        val sessHeaders = arrayOf(context.getString(R.string.pdf_h_provider), context.getString(R.string.pdf_h_project), context.getString(R.string.pdf_h_cost), context.getString(R.string.pdf_h_usage), context.getString(R.string.pdf_h_status))
        for (i in sessHeaders.indices) {
            canvas.drawText(sessHeaders[i], MARGIN + CONTENT_WIDTH * sessCols[i], y + 9f, smallBoldPaint)
        }
        y += 14f
        divider(thin = true)

        val topSessions = sessions.sortedByDescending { it.estimatedCost }.take(15)
        for (s in topSessions) {
            checkSpace(14f)
            val row = arrayOf(
                s.provider.take(18),
                s.project.take(18),
                "$${String.format("%.4f", s.estimatedCost)}",
                formatTokens(s.totalUsage),
                (DataTokenDisplay.sessionStatus(s.status)?.let { context.getString(it) } ?: s.status).take(10),
            )
            for (i in row.indices) {
                canvas.drawText(row[i], MARGIN + CONTENT_WIDTH * sessCols[i], y + 9f, smallPaint)
            }
            y += 13f
        }
        y += 8f

        // Daily cost trend
        if (dailyUsage.isNotEmpty()) {
            checkSpace(40f)
            divider()
            y += 4f
            val costByDate = dailyUsage.groupBy { it.date }.mapValues { (_, items) -> items.sumOf { it.cost } }
            val sortedDates = costByDate.keys.sorted().takeLast(30)
            // The heading said "Last 30 Days" whatever range the shared repository held
            // (7 or 14 after a visit to Cost Analysis). State the span actually plotted.
            val spanDays = trendSpanDays(sortedDates)
            canvas.drawText(
                context.resources.getQuantityString(R.plurals.pdf_daily_trend_days, spanDays, spanDays),
                MARGIN, y + 16f, headingPaint,
            )
            y += 24f

            val maxCost = costByDate.values.maxOrNull() ?: 1.0

            for (date in sortedDates) {
                checkSpace(13f)
                val cost = costByDate[date] ?: 0.0
                val barWidth = if (maxCost > 0) (cost / maxCost * (CONTENT_WIDTH - 130)).toFloat() else 0f

                canvas.drawText(DateDisplay.monthDay(date, monthDayPattern, locale) ?: date.takeLast(5), MARGIN, y + 8f, labelPaint)
                canvas.drawRect(MARGIN + 45f, y, MARGIN + 45f + barWidth, y + 8f, barPaint)
                canvas.drawText("$${String.format(Locale.ROOT, "%.2f", cost)}", MARGIN + 50f + CONTENT_WIDTH - 130f, y + 8f, labelPaint)
                y += 12f
            }
        }

        // Footer
        checkSpace(30f)
        y += 10f
        divider()
        canvas.drawText(context.getString(R.string.pdf_footer, BuildConfig.VERSION_NAME, today), MARGIN, y + 8f, labelPaint)

        doc.finishPage(page)

        // Write to file
        val exportDir = File(context.cacheDir, "cli_pulse_exports").also { it.mkdirs() }
        val file = File(exportDir, "cli-pulse-report-${dateFormat.format(Date())}.pdf")
        return try {
            FileOutputStream(file).use { doc.writeTo(it) }
            doc.close()
            file
        } catch (_: Exception) {
            doc.close()
            null
        }
    }

    private fun formatTokens(value: Int): String = when {
        value >= 1_000_000 -> String.format("%.1fM", value / 1_000_000.0)
        value >= 1_000 -> String.format("%.1fK", value / 1_000.0)
        else -> "$value"
    }

    /** Calendar days from the first to the last plotted date, inclusive; the count when dates do not parse. */
    internal fun trendSpanDays(sortedDates: List<String>): Int {
        if (sortedDates.isEmpty()) return 0
        return runCatching {
            java.time.temporal.ChronoUnit.DAYS.between(
                java.time.LocalDate.parse(sortedDates.first()),
                java.time.LocalDate.parse(sortedDates.last()),
            ).toInt() + 1
        }.getOrDefault(sortedDates.size)
    }
}
